deckhouse-controller/pkg/helpers/change_registry/change_registry.go

```diff
@@ -0,0 +1,367 @@
+		return err
+	}
+	return nil
+}
+
+func updateImageRepoForContainers(containers []v1.Container, newRepository string, nameOpts []name.Option, remoteOpts []remote.Option, newTagsForContainer map[string]string /* map[<container name>]<newTag> */) ([]v1.Container, error) {
+	for i, container := range containers {
+		oldImage, err := name.ParseReference(container.Image)
+		if err != nil {
+			return nil, err
+		}
+
+		tagOrDigestref := oldImage.Identifier()
+		if newTagsForContainer != nil {
+			if v, f := newTagsForContainer[container.Name]; f && v != "" {
+				tagOrDigestref = v
+			}
+		}
+
+		var delim string
+		switch oldImage.(type) {
+		case name.Digest:
+			delim = "@"
+		case name.Tag:
+			delim = ":"
+		}
+
+		newRef, err := name.ParseReference(newRepository+delim+tagOrDigestref, nameOpts...)
+		if err != nil {
+			return nil, err
+		}
+
+		// Check that new reference exists in the new repo before applying changes.
+		if err := checkImageExists(newRef, remoteOpts); err != nil {
+			return nil, fmt.Errorf("\nimage '%s' doesn't exist: %w", newRef.Name(), err)
+		}
+
+		container.Image = newRef.Name()
+		containers[i] = container
+	}
+
+	return containers, nil
+}
+
+func checkImageExists(imageRef name.Reference, opts []remote.Option) error {
+	img, err := remote.Image(imageRef, opts...)
+	if err != nil {
+		return err
+	}
+
+	if _, err := img.Digest(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// checkBearerSupport func checks that registry accepts bearer token authentification.
+// This is modified "ping" func from
+// https://github.com/google/go-containerregistry/blob/v0.5.1/pkg/v1/remote/transport/ping.go
+func checkBearerSupport(ctx context.Context, reg name.Registry) error {
+	const (
+		bearer        = "bearer"
+		wwwAuthHeader = "WWW-Authenticate"
+	)
+
+	client := http.Client{Transport: http.DefaultTransport}
+
+	// This first attempts to use "https" for every request, falling back to http
+	// if the registry matches our localhost heuristic or if it is intentionally
+	// set to insecure via name.NewInsecureRegistry.
+	schemes := []string{"https"}
+	if reg.Scheme() == "http" {
+		schemes = append(schemes, "http")
+	}
+
+	var errs []string
+	for _, scheme := range schemes {
+		url := fmt.Sprintf("%s://%s/v2/", scheme, reg.Name())
+		req, err := http.NewRequest(http.MethodGet, url, nil)
+		if err != nil {
+			return err
+		}
+		resp, err := client.Do(req.WithContext(ctx))
+		if err != nil {
+			errs = append(errs, err.Error())
+			// Potentially retry with http.
+			continue
+		}
+		defer func() {
+			// By draining the body, make sure to reuse the connection made by
+			// the ping for the following access to the registry
+			defer resp.Body.Close()
+			io.Copy(io.Discard, resp.Body)
+		}()
+
+		if resp.StatusCode != http.StatusUnauthorized {
+			return transport.CheckError(resp, http.StatusUnauthorized)
+		}
+
+		if challenges := authchallenge.ResponseChallenges(resp); len(challenges) != 0 {
```

Comment:
Let's avoid unnecessary docker distribution dependency. It should be enough to just copy the `ResponseChallenges` method.